tensorflow/python/keras/utils/metrics_utils.py

```diff
@@ -424,7 +424,7 @@ def update_confusion_matrix_variables(variables_to_update,
 
   def weighted_assign_add(label, pred, weights, var):
     label_and_pred = math_ops.cast(
-        math_ops.logical_and(label, pred), dtype=dtypes.float32)
+        math_ops.logical_and(label, pred), dtype=pred.dtype)
```

Comment:
This doesn't look right.

`logical_and` only works if both inputs are dtype `bool`, it returns tensor with dtype `bool`. so here you're casting a bool to type bool.

It needs to be cast to a floating type so the rest of the function works (`*` and `reduce_sum` don't work on bool).